Source/ImageFetcher.swift

```diff
@@ -1,82 +1,74 @@
 import UIKit
 import Foundation
 
-public class ImageFetcher {
+struct ImageFetcher {
 
-  public enum Result {
+  enum Result {
     case Success(UIImage)
     case Failure(ErrorType)
   }
 
-  public enum Error: ErrorType {
+  enum Error: ErrorType {
     case InvalidResponse
     case InvalidStatusCode
     case InvalidData
     case ConvertionError
   }
 
-  public let URL: NSURL
-  var task: NSURLSessionDataTask?
-  var active = false
-
-  public var session: NSURLSession {
+  static var session: NSURLSession {
     return NSURLSession.sharedSession()
   }
 
-  // MARK: - Initialization
-
-  public init(URL: NSURL) {
-    self.URL = URL
-  }
+  static var imageAddresses = [String: NSURLSessionDataTask]()
 
   // MARK: - Fetching
 
-  public func start(completion: (result: Result) -> Void) {
-    active = true
-
-    task = session.dataTaskWithURL(URL) { [weak self] data, response, error -> Void in
-      guard let weakSelf = self where weakSelf.active else { return }
-
-      if let error = error {
-        weakSelf.complete { completion(result: .Failure(error)) }
-        return
-      }
-
-      guard let httpResponse = response as? NSHTTPURLResponse else {
-        weakSelf.complete { completion(result: .Failure(Error.InvalidResponse)) }
-        return
-      }
-
-      guard httpResponse.statusCode == 200 else {
-        weakSelf.complete { completion(result: .Failure(Error.InvalidStatusCode)) }
-        return
-      }
-
-      guard let data = data where httpResponse.validateLength(data) else {
-        weakSelf.complete { completion(result: .Failure(Error.InvalidStatusCode)) }
-        return
+  static func fetch(URL: NSURL, imageAddress: String, completion: (result: Result) -> Void) {
+    dispatch_async(dispatch_get_main_queue()) {
+      if let task = imageAddresses[imageAddress] {
+        task.cancel()
       }
 
-      guard let image = UIImage.decode(data) else {
-        weakSelf.complete { completion(result: .Failure(Error.ConvertionError)) }
-        return
+      let task = session.dataTaskWithURL(URL) { data, response, error -> Void in
+        if let error = error {
+          complete() { completion(result: .Failure(error)) }
+          return
+        }
+
+        guard let httpResponse = response as? NSHTTPURLResponse else {
+          complete(imageAddress) { completion(result: .Failure(Error.InvalidResponse)) }
+          return
+        }
+
+        guard httpResponse.statusCode == 200 else {
+          complete(imageAddress) { completion(result: .Failure(Error.InvalidStatusCode)) }
+          return
+        }
+
+        guard let data = data where httpResponse.validateLength(data) else {
+          complete(imageAddress) { completion(result: .Failure(Error.InvalidStatusCode)) }
+          return
+        }
+
+        guard let image = UIImage.decode(data) else {
```

Comment:
I checked that, the thing is that the task resolves in a random thread every time, so we are good.